---
title: "Homework 3. data table"
author:
date: "`r format(Sys.time(), '%d.%m.%Y')`"
output: html_notebook
---


# Deadline for homework is Sunday,24.10.2021., 23.55 h

# Instructions and RULES:
+ Perform all of the analysis using only data.table. 
+ It is allowed to use functions that are present in base R (DO NOT forget them as you may need to search them and use them in this homework).
+ All the solutions which use for loops will not be graded. Use the functions from the apply family!!!
+ If certain task contains detailed instructions how to solve something, then those instructions must be followed or these tasks will not be graded.
+ Whenever you are asked to show a certain value or values as an answer and if it is not precisely stated in which format should the answer be (data.table or a vector) then both answers are correct.
+ Manual calculations and writing the answers directly on the code chunk will not be taken into consideration without the proper coding steps (and/or saved variable) that confirm this answer.


```{r}
library(data.table)
```

# Task 1. Analysis of dataset *flights* (44 pts)

Read in the flights data in flights variable from   "https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv". (Install curl package if you need to before doing this.)

```{r, message=FALSE}
library(data.table)
flights <- fread("https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv")
flights
```

a1) Order the flights data table by carrier (ascenting order) then by decreasing distance, and save only hour as a vector in a variable var1a1. (1 point)
```{r}
var1a1<-flights[order(carrier, -distance), hour]; head(var1a1)
```

a2) Order the flights data table by carrier (ascenting order) then by decreasing distance, and save only hour as a data table with one column in a variable var1a2. (1 point)
```{r}
var1a2<-flights[order(carrier, -distance), -(year:distance)]; var1a2
```

b1) Find the max flight air time for each carrier. copy the result to variable var1b1 (data.table with 2 columns, carrier and maxairtime, sorted by carrier. Do this in one command, no chaining, don't change the original data.table.). (1 point)  
```{r}
var1b1<-flights[order(carrier), .(maxairtime = max(air_time)), by=carrier]; var1b1
```
b2) Find the max flight air time for each carrier. copy the result to variable var1b2 (data.table with 2 columns, carrier and maxairtime, sorted by maxairtime, from largest to smallest. Don't change the original data.table.). (1 point)  
```{r}
var1b2<-flights[, .(maxairtime = max(air_time)), by=carrier][order(-maxairtime)]; var1b2
```
c1) Count the number of flights that flew maximum air time for each carrier. Copy the data.table with carrier and numberofmaxflights to var1c1. Do this in one command, no addition of new columns and no chaining=2 points.  
```{r}


var1c1<-flights[, .N, by=.(carrier, air_time==max(air_time))];var1c1


#var1c1<-flights[,.(numberofmaxflights = {
#  maxi=max(air_time)
  
#  }), by=.(carrier)]

#Do not know how to solve this task
```
c2) Then do this by addition of new column to flights and/or chaining in one command = 1 point save the result to var1c2 (copy!). If you add the column, remove it in the end of this exercise (end of c2 just remove the column/s if added from flights).  
```{r}
var1c2<-flights[, maxairtime:= max(air_time), by=carrier][air_time==maxairtime, .(numberofmaxflights = .N), by=carrier]; var1c2

#removal of column added
flights[, maxairtime := NULL]
flights
```
d1) Count how many flights had the same flight time for each carrier separately, ordered by carrier alphabetically and by decreasing flight time (carrier, air_time, N). copy the result to var1d1. Do this in one command, no chaining. (1 point)
```{r}
var1d1<-flights[order(carrier, -air_time), .N, by=.(carrier, air_time)]; var1d1
```

d2) Get the data.table containing the same as previous (carrier, air_time, N), but only for the longest 2 flights for each carrier! copy the result to var1d2. Do this in one command, chaining if needed. (1 point)  
```{r}
#two options
var1d2<-flights[order(carrier, -air_time), .SD[c(1, 2)], .SDcols=c("air_time"), by=carrier][, .N, by=.(carrier, air_time)]; var1d2
#var1d2<-flights[order(carrier, -air_time), .(N = .N), by=.(carrier, air_time)][, .SD[c(1, 2)], .SDcols=c("air_time", "N"), by=carrier]; var1d2
```

d3) Get the data.table containing the same as previous (carrier, air_time, N), but only for the longest and shortest flights for each carrier! copy the result to var1d3. Do this in one command, chaining if needed. (1 point)  
```{r}
var1d3<-flights[order(carrier, -air_time), .(N = .N), by=.(carrier, air_time)][, .SD[c(1, .N)], .SDcols=c("air_time", "N"), by=carrier]; var1d3
```
e) Get the name of the carrier who has the largest difference between longest and shortest flight air time! copy the result to var1e (only the carrier name as an element of vector). Do this in one   command, chaining if needed, do not change the original data.table! (3 points)  
```{r}
var1e<-flights[, .(difference = max(air_time) -  min(air_time)), by=carrier][, carrier[difference == max(difference)]]; var1e
```
f) Get the name of the carrier who has the largest mean air time. One command, chaining allowed, don't change the original data.table. Save the name alone in variable var1f (still the same command.). (1 point)  
```{r}
var1f<-flights[, .(meanairtime = mean(air_time)), by=carrier][, carrier[meanairtime == max(meanairtime)]]; var1f
```
g1) Write a function that will calculate the squared sum of differences between each element and the mean value of all elements, call the function myRss. The function should take one vector as an argument and return numeric value. Use this function to calculate the value for air time for each carrier, name the column rss, dont change the original table! one command, no chaining. (copy to var1g1, 2 points).  
```{r}
myRss<-function(vec1){
  sum(sapply(vec1, function(v) (v - mean(vec1))**2))
}

var1g1<-flights[, .(rss = myRss(air_time)), by=carrier]; var1g1
```
g2) Do as the previous but for air_time and arr_delay columns in the same command, no chaining, no change to original data.table (var1g2, 1 point)!   
```{r}
var1g2<-flights[, lapply(.SD, myRss), .SDcols=c("air_time", "arr_delay"), by=carrier]; var1g2
```
g3) Do as the previous but for all numeric columns in the same command, no chaining, no change to original data.table (var1g3, 2 points)!   
```{r}
var1g3<-flights[, lapply(.SD, myRss), .SDcols = is.numeric, by=carrier];var1g3
```
h1) Assign a new column to flights: speed which will have speed calculated for each flight. Do it in a data.table fashion. copy the resulting flights data.table to myFlights1 variable. (1 point)  
```{r}
flights[, speed := (distance/air_time)]
myFlights1<-copy(flights); myFlights1
```
h2) Convert speed and distance variables to type double. do it in data.table fashion, in one command. copy the resulting data.table to myFlights2 variable. Do this withouth lapply. (1 point). 
```{r}
flights[, ":=" (speed = as.double(speed),
                   distance = as.double(distance))]; flights

myFlights2<-copy(flights); myFlights2
```
h3) Again do the RSS calculation as in var1g1 but for all double columns in the same command, no chaining, no change to original data.table (var1g4, 1 point)!   
```{r}
var1g4<-myFlights2[, lapply(.SD, myRss), .SDcols = is.double, by=carrier]; var1g4
```
i) Save the names of the columns which are of type double to thosecolumns variable.(1 point)
```{r}
thosecolumns<-myFlights2[, names(.SD), .SDcols = is.double]; thosecolumns
```
j) Convert thosecolumns back to type integer. do it in data.table fashion, in one command.  copy the resulting data.table to myFlights3 variable. But now do this with lapply.(2 points).
```{r}
flights[, c(thosecolumns) := lapply(.SD, as.integer), .SDcols=thosecolumns]

myFlights3<-copy(flights); myFlights3
```

k) For air_time column, calculate the difference between the second and first observation for each carrier, in flights data frame ordered by carrier and then by decreasing air_time. In the same command, no chaining, no change to original data.table, no additional function defined outside (var1k, 2 points)!
```{r}
var1k<-flights[order(carrier, -air_time), c(dif = .SD[2] - .SD[1]), .SDcols="air_time", by = carrier]; var1k
```
l) For all integer columns, calculate the difference between the second and first observation for each carrier. In the same command, no chaining, no change to original data.table, no additional function defined outside (var1l, 2 points)!
```{r}
var1l<-flights[, c(dif = .SD[2] - .SD[1]), .SDcols=is.integer, by = carrier]; var1l
```
m) Copy the flights to myFlights. Add a new column, rss_air_time which will have myRss of air_time saved in it, to myFlights. do not change the flights variable, dont use chaining, do it in one command, (myFlights assignment is separate command.) copy the result to variable myFlights1
```{r}
myFlights<-copy(flights)
myFlights1<-myFlights[, rss_air_time := myRss(air_time)]; myFlights1

```
n) Make a vector with all the names of the columns with integer values, save it in intnames variable. Use the **paste** function to make a vector which will have all the names of integer columns converted to rss_columnname (so for air_time it will be rss_air_time, dor hour it will be rss_hour, ...) (2 points) (No need to do this with data table). save the vector as rssnames variable.
```{r}

intnames<-myFlights[, names(.SD), .SDcols = is.integer]; intnames
rss_columname<-paste("rss", intnames, sep="_"); rss_columname

#I am assuming I am working with myFlights as a copy of the version I have now of flights: with speed column added and that is type integer -because when I created in task h1 it was type double but at the end of task j it is type integer-.
```
o) Use the previously created vectors to add multiple new columns to data.table. The names of the columns you will add will be rssnames. And the values will of course be myRss(column), where column is every column from intnames. Do it in one command, data.table style, no chaining. Copy the result to myFlights3 variable. (2 points).
```{r}

myFlights3<-myFlights[, c(rss_columname) := lapply (.SD, myRss), .SDcols=(intnames)]; myFlights3

```
p) Go back to flights data.table. Calculate and return (do not change the data.table) the mean, and sd for air_time for flights which travel between JFK and MIA. Do this by each carrier, in one line, no chaining. Do not rename the columns to anything, copy the result to myFlights4.(1 points)  
```{r}
#myFlights4<-flights[(origin=="JFK" & dest=="MIA"), .(mean_JFK_MIA = mean(air_time),
#                                                      sd_JFK_MIA = sd(air_time)),
#                     by=carrier]; myFlights4

#I do not know if by not renaming the columns you refer to those that have the values I just calculated so I will leave both options (first with colum names, second without)

myFlights4<-flights[(origin=="JFK" & dest=="MIA"), .(mean(air_time),
                                                    sd(air_time)),
                     by=carrier]; myFlights4

#I have not specified MIA->JFK travels too (origin=="MIA" & dest="JFK") because I checked and there are any flights with that characteristics
```
r)  Which journey (from -> to, regardless of the carrier) was delayed in arrival the most times? (var1r, 1 pts)
```{r}
var1r<-flights[arr_delay > 0, .(N = .N), by=.(origin, dest)][N==max(N), .(origin, dest)]; var1r
```
s)  Which journey (from -> to, regardless of the carrier) was delayed in arrival the most in total minutes? (var1s, 1 pts)
```{r}
var1s<-flights[, .(total = sum(arr_delay)), by=.(origin, dest)][total==max(total), .(origin, dest)]; var1s
```
t) Are flights longer depending on time of the day? (Divide the day into : night (from 22->6 hours), morning (6->12, afternoon 12->16, evening
16->22 )) (var1t, 3)
```{r}
day_time<-function(h){
  if (h >= 22 | h < 6){
    time_of_day<-"night"
    } 
  else if (h >= 6 & h < 12){
    time_of_day<-"morning"
    }
  else if (h >= 12 & h < 16){
    time_of_day<-"afternoon"
    }
  else {time_of_day<-"evening"}
  }
  
var1t<-flights[, time_of_the_day := sapply(hour, day_time)][, .(mean_by_daytime = mean(air_time)), by=time_of_the_day]; var1t

#Regarding the obtained results we can see that flights on mornings and evenings tend to be longer than those that departure in the afternoon, being the ones that departure at night usually the shortest ones.
```
u) For which time of the day are flights delayed in departure the most and for which the least? What about arrival delays? (var1u, 2)
```{r}
var1u<-flights[, .(mean_of_dep = mean(dep_delay),
                   mean_of_arr = mean(arr_delay)), by = time_of_the_day][, .(dep_most_delayed = time_of_the_day[mean_of_dep == max(mean_of_dep)],
                                                                           dep_least_delayed = time_of_the_day[mean_of_dep == min(mean_of_dep)],
                                                                           arr_most_delayed = time_of_the_day[mean_of_arr == max(mean_of_arr)],
                                                                           arr_least_delayed = time_of_the_day[mean_of_arr == min(mean_of_arr)])]
var1u

#The flights that departure and arrive in the night are the most delayed and the ones that do so in the mornings are the least delayed.
```
v) Answer u) but take into account the day of the week as well. Use package function **wday** from package **lubridate** to answer this question. (For which day of the week and which time of the day are flights delayed the least, and for which the most?) (var1u, 2)
```{r}

#First I create a vector with the names for the days of the week in order to subset the result of using the wday function in flights datatable. It is important to note that wdays starts on Sunday, so number 1 would be Sundays, 2 would be Mondays,... an so on. Then I crete a new column in flights datatable with the weekday for each column and finally I created new (not permanent) columns with the answers of the questions (obtained by subsetting on the flights datatable). Note that first row is the name of the day and the second row is the period of the day.

Daysoftheweek<-c("Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday")

flights[, day_of_week := 
          {number = wday(mapply(paste, year, month, day, sep ="-"))
          Daysoftheweek[number]}]


var1v<-flights[, .(mean_of_dep = mean(dep_delay), mean_of_arr = mean(arr_delay)), by = .(day_of_week,  time_of_the_day)]

var1v<-var1v[, .(dep_most_delayed = c(day_of_week[mean_of_dep == max(mean_of_dep)], time_of_the_day[mean_of_dep == max(mean_of_dep)]),
              dep_least_delayed = c(day_of_week[mean_of_dep == min(mean_of_dep)], time_of_the_day[mean_of_dep == min(mean_of_dep)]),
              arr_most_delayed = c(day_of_week[mean_of_arr == max(mean_of_arr)], time_of_the_day[mean_of_arr == max(mean_of_arr)]),
             arr_least_delayed = c(day_of_week[mean_of_arr == min(mean_of_arr)], time_of_the_day[mean_of_arr == min(mean_of_arr)]))]
var1v

#The flights that take off and land on Wednesday nights are the most delayed and the ones that do so on Sunday mornings are the least delayed

#I am sorry for the "salami code" but it is the only way I know for saving all the answers for the questions in the same variable.
```


# Task 2. Taxonomic analysis of sponge genomes (25 pts)

## Short introduction
Program DIAMOND is a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data. When combined with program MEGAN, it becomes an indispensable tool for taxonomic and functional analysis of metagenomes. More information can be found at https://link.springer.com/protocol/10.1007/978-1-4939-9074-0_19. Not only is it useful for analyzing metagenomes but this workflow can also be used in detection of contaminants in newly assembled genomes of Eukaryotes. In this task you will have to analyze DIAMOND results for the genomes of several sponges.

Before you start solving this task, you need to understand how BLAST and DIAMOND work since you are analyzing its output.
Answer these questions before solving the tasks. (1)

What is DIAMOND and why is it similar to BLAST? 
#DIAMOND is a high-throughput program (a tool) that enables us to make alignments of protein sequences or translated DNA reads against a protein reference database. Is it mostly the same as blastp or blastx but it performs the aligns up to 20,000 times the speed of BLAST.

What is the output format and what do the columns represent (focus on columns present in table *spongeDiamond* in step 0. in the following task)?
#The output is a DIAMOND alignment archive (DAA), is a proprietary binary format that can be used to generate other output formats. It can be supported by MEGAN and allows quick import of the results.The columns represent: sponge is the genre of sponge, qseqid is the ID of the query sequence (query sequence being the protein of translated DNA sequence you want to align), qlen is the lenght of the query sequence, the sseqid is the ID of the subject sequence, the slen is the lenght of the subject sequence, staxids is taxonomy ID and stitle is the subject title. Qcovhsp is the Query Coverage per HSP, lenght is the lenght of the alignment, pident is the percentage of identical matches, missmatch is the number of identical matches, qstart is the start of the alignment in query, qend is the end of the alignemnt in query, sstart and send are the same but for subject sequence. Evalue is the expected value (the number of expected hits of similar quality (score) that could be found just by chance) what means that the smaller the evalue the most reliable is the alignment. And bitscore is the Bitscore, another statistical indicator (such as evalue) that measures the sequence similarity independently of query length and database size. The rest, from superkingdom to species are different levels of taxons.

terminar

## Tasks
0.step: Import the file **SpongeDiamond_algorithms2021.RDS** and examine the table.
```{r}
spongeDiamond <- readRDS("Task2/SpongeDiamond_algorithms2021.RDS")
spongeDiamond
```

a) How many sponges do you have in your table? (1)
```{r}
var2a<-spongeDiamond[,uniqueN(sponge)]; var2a
```
b) Which sponge (or sponges if more have the same result) has the most contigs? (1)
```{r}
var2b<-spongeDiamond[, .(dif_contig = uniqueN(qseqid)), by=sponge][dif_contig == max(dif_contig), sponge]; var2b
```
c) Which sponge (or sponges if more have the same result) has the shortest median length of contigs? (1)
```{r}
var2c<-spongeDiamond[, .(median_len = median(uniqueN(qlen))), by=sponge][median_len==min(median_len), sponge]; var2c
```
d) Which contig has the highest and which contig the lowest number of mismatches in total for each sponge? (1) 
```{r}
##revisar

var2d<-spongeDiamond[, .(total_mismatch = sum(mismatch)), by=.(sponge, qseqid)][, c(qseqid[total_mismatch = max(total_mismatch)], qseqid[total_mismatch = min(total_mismatch)]), by=sponge]; var2d

#I do not know why this does not work, I think is because there are more than 1 contig for each sponge that have a total number mismatches equal to 0 so there are plenty of contigs for each sponge that have the lowest number.

```
e) Which pair of superkingdom-family occurred most time for Amphimedon? (2)
```{r}
var2e<-spongeDiamond[, .(N=.N), by = .(sponge=="Amphimedon", superkingdom, family)][, c(superkingdom[N==max(N)], family[N==max(N)])]; var2e
```
f) Select protein-sponge pair with the highest aligned lengths for each pair and calculate the difference between the pairs with the highest sum of length and lowest sum of lengths. (2)
```{r}
#Protein-sponge pairs with highest aligned lengths for each pair
var2f1<-spongeDiamond[, max(length), by=.(sponge, sseqid)]; var2f1

#difference between the pairs with the highest sum of length and lowest sum of lengths. I am not sure if the task is to calculate the difference between the highest and lowest lengths of the pairs with the highest aligned lengths for each pair (the ones in var2f1) or to calculate the difference of those pairs that have the highest and lowest sum of lengths. Just in case, here it is how I would calculate either possibility:

var2f21<-spongeDiamond[, .(max_len = max(length)), by=.(sponge, sseqid)][, max(max_len) - min(max_len)]; var2f21
var2f22<-spongeDiamond[, .(sum_len = sum(length)), by=.(sponge, sseqid)][, max(sum_len) - min(sum_len)]; var2f22
```
g) Annotate every query contigs for each sponge in the original cleaned *spongeDiamond* table in the following way: If best 3 results (by eval) for each contig belongs to the same superkingdom, annotate that contig as that superkingdom. (2) Note: Select smaller sample size on which you will develop your code and then run it on the whole dataset if you are having troble with the running time or crashing. 
```{r}
sup_king_anot2<-function(v){
  if (is.na(v[1]) || is.na(v[2]) || is.na(v[3])){
    anot<-"NA"
  }
  else if (v[1] == v[2] && v[2] == v[3]){
    anot<-v[1]
  }
  else {
    anot<-"NA"
  }
  anot
}

var2g<-spongeDiamond[order(evalue), .SD[c(1:3)], .SDcols = c("superkingdom"), by=.(sponge, qseqid)][, anot := sup_king_anot2(superkingdom),  by=.(sponge, qseqid)]; var2g

```
  + What is the difference in the number of contaminants (sequences belonging to superkingdom Bacteria) between the sponge with the highest number of contaminants and the rest ? Show the results as data table with sponges and differences (3)
```{r}
max_value<-spongeDiamond[superkingdom=="Bacteria", .(N = .N), by=.(sponge)][,max(N)];
var2g1<-spongeDiamond[superkingdom=="Bacteria", .(diff = {
  N = .N
  max_value - N}), by=sponge]
var2g1
```

h) Add a column "pidentGrp" that will represent identity groups for each row : cut the “pident” value to bins of 0-10%, 10-20%, 20-30%,…90-100%. 
Add a column "maxLength", that will represent maximum of aligned length for each contig in sponge- this is the largest match possible for each contigs and all hits. Add another column "percMaxLen" that will say what percentage of maxLength the row has. (3)
```{r}
percent<-function(v){
  first<-floor(v/10)
  ifelse(first==10, 9, first)
  second<-as.integer(first)+1
  res<-paste(paste(first, 0, sep = ""), paste(first+1, 0, "%", sep = ""), sep = "-")
  res
}

spongeDiamond[, maxLength := max(length), by=.(sponge, qseqid)]
spongeDiamond[,  ":=" (percMaxLenth = (length/maxLength) * 100,
                       pindentGrp = percent(pident))]

spongeDiamond

#I do not know why the ifelse condition in function percent does not work so those rows with pident==100.0 would be in the 90-100% interval and not in the 100-110%.
```
i) In this part functions rleid and shift might be useful to you. What do they do? (2 points). 
```{r}
#Rleid is a function that generates a run-length type id column to be used in grouping operators. It is useful when you need to do some grouping such as when you need to perform operations where consecutive runs of identical values should be part of the same group. It assigns an id when a new value appears and assigns the same id to the following rows if their value is the same as the previous one. When a new different value appears it will assign a new id (the next in row, 2 for example)  and does the same. When some value that has already appeared shows up again it will assing a new id (NOT the same as before). 

#shift is a fast lead lag for vectors and lists (also dataframes and datatables).It always returns a list except when the input is a vector and length is 1 (a vector is obtained). This functions shifts a vector the number of rows you specify and creates lag or log variables.
```
  + Read in file with one whole contig for Tethya (**contigTethya.txt**) find which "class" appears most times. (save the Tethya genome to new a variable "genomeTethya" and continue to work with it). For example if "class" repeat is Demospongiae, Demospongiae, Leptocardii, Leptocardii, Leptocardii, Leptocardii then the class Leptocardii appears most times consecutively, 3 time to be precise. If multiple class appear the same times, report all of them. (4 pts)
```{r}
genomeTethya <- data.table::fread("Task2/contigTethya.txt")

var2i1<-genomeTethya[, consec_times := (.N - 1), by=rleid(class)][, unique(class[consec_times==max(consec_times)])];
var2i1
```

  +  In which positions are those Families (do the analysis on the variable "longestTethya")? Save the result as vector. (2 points)
```{r}

longestTethya<-copy(genomeTethya)
longestTethya[,  I := .I]
var2i1<-longestTethya[, I[consec_times==max(consec_times)]]; var2i1

#I think the answer I gave is not correct as it does not refer to anything related to families. I think what I am supposed to do is to create a list with the positions (row numbers) for the most consecutively repeated classes by family, so I should I associate every family to several positions. I think I should do it with the shift function (as I saw in some bibliography that this command can associate some names to some values of a data.table... or something like that). I really do not understand the syntax of this function and I tried several options but I don't think I can reach an answer for this matter. Here are some examples of what I was trying to but I think it would be very helpful to see how this function works as I only found more confusing info on the internet. 

#trial1<-longestTethya[, where := shift(.SD), .SDcols = c("family", "I"), by=consec_times]; trial1
#trial2<-longestTethya[, shift(family, I), by=(consec_times==max(consec_times))]; trial2

```


