---
title: "Homework 4. Data manipulation"
author: "Lucía Muñoz Gil"
output: html_notebook
---

# 91% (40,75/45)

# Deadline for homework is Sunday, 30.10.2021., 23.55 h

# Instructions and RULES:
+ Perform all of the analysis using only data.table. 
+ It is allowed to use functions that are present in base R (DO NOT forget them as you may need to search them and use them in this homework).
+ All the solutions which use for loops will not be graded. Use the functions from the apply family!!!
+ If certain task contains detailed instructions how to solve something, then those instructions must be followed or these tasks will not be graded.
+ Whenever you are asked to show a certain value or values as an answer and if it is not precisely stated in which format should the answer be (data.table or a vector) then both answers are correct.
+ Manual calculations and writing the answers directly on the code chunk will not be taken into consideration without the proper coding steps (and/or saved variable) that confirm this answer.


# Task 1. NMR peaks (10 pts)
Goal: In this task you will have to upload multiple files in R and save certain calculation in a newly created format. You will be working with peak lists and intensity files for urine samples measured by 1H NMR (Psihogios NG, et al.). Files are located under the folder Task1.

a) Import the files located in folder Task1. Make one table of all the imported files. The table must also have sample id (only the name of the file) and group name (severe kidney disease and healthy). (Since you have not learned regular expression and string manipulation, solve this task without it.). **Do not read files one at the time! Hard coding will not be evaluated!** (5)
```{r}

temp1<-list.files(path="../Task1/Healthy", full.names = TRUE)
temp_hel<-list.files(path="../Task1/Healthy")

temp2<-list.files(path="../Task1/Severe_kidney_disease", full.names = TRUE)
temp_kid<-list.files(path="../Task1/Severe_kidney_disease")

one<-rbindlist(lapply(setNames(temp1, temp_hel), fread), idcol=TRUE)
one[, Group := "Healthy"]

dos<-rbindlist(lapply(setNames(temp2, temp_kid), fread), idcol=TRUE)
dos[, Group := "Severe Kidney Disease"]

var1a<-rbind(one, dos); var1a

```
<span style="color:green"> 5/5 pts great, now try to think of a ways to make u function to reduce the repeating of import. </span>

b) Crate a new subdirectory under the directory "Task1" called "results" using R. Make an R object that contains two tables. You are going to save the R object in the created directory "results". The first table has the summary statistic of ppm of each group (sum, mean, median, min, max). While the second table contains the same summary statistics of ppm but only for three patient with severe kidney disease that have the highest overall intensity. The calculated results must be in rows and groups/patients in columns. (5)
```{r}
#creation of subdirectory results
dir.create("../Task1/results")

#First table: sumary statistic of ppm of eeach group
fst_table<-var1a[, .(sum = sum(ppm),
                     mean = mean(ppm),
                     median = median(ppm),
                     min = min(ppm),
                     max = max(ppm)), by=Group]; fst_table

fst_table1<-as.data.table(t(fst_table)[-1, ])
colnames(fst_table1)<-fst_table[, Group]
rownames(fst_table1)<-colnames(fst_table)[-1]
fst_table1

#Second table
three_highest<-var1a[Group=="Severe Kidney Disease", .(int_mean = mean(INTENSITY)), by=.id][order(-int_mean), .id[1:3]]

sec_table<-rbindlist(lapply(three_highest, function (x) var1a[.id==x, .(sum = sum(ppm),
                                                                        mean = mean(ppm),
                                                                        median = median(ppm),
                                                                        min = min(ppm),
                                                                        max = max(ppm))])); sec_table

sec_table1<-as.data.table(t(sec_table))
colnames(sec_table1)<-three_highest
rownames(sec_table1)<-colnames(sec_table)
sec_table1

#Saving the R object
save(fst_table1, sec_table1, file = "../Task1/results/twotables.RData")
```
<span style="color:green"> 5/5 pts. </span>

# Task 2. GC content sponges (20 pts)
Table mononucleotides.txt and dinucleotides.txt contain mono/dinucleotide counts in 7 different sponge species. Use all data.table operations as you need them in these task. 
Your tasks are as follows:
```{r}
mononucleotides <- data.table::fread("http://hex.bioinfo.hr/~mfabijanic/mononucleotides.txt")
dinucleotides <- data.table::fread("http://hex.bioinfo.hr/~mfabijanic/dinucleotides.txt")
```

(3 pts) - Change name of the first column to "species". Calculate mono/dinucleotide frequencies for each species (do this with and without *melt* function and in later tasks proceed with only one table of your choice).
```{r}
#CHENGE OF NAME
lapply(list(mononucleotides, dinucleotides), function (x) setnames(x, "rn", "species"))

#FRECUENCY CALCULATION WITHOUT MELT
mono_col<-colnames(mononucleotides)[-1]
freq_mono<-copy(mononucleotides)
freq_mono<-freq_mono[, lapply(.SD, function (x) x/sum(.SD)), .SDcols=mono_col, by=species]; freq_mono

di_col<-colnames(dinucleotides)[-1]
freq_di<-copy(dinucleotides)
freq_di<-freq_di[,lapply(.SD, function (x) x/sum(.SD)), .SDcols=di_col, by=species]; freq_di

#FRECUENCY CALCULATION WITH MELT
freq_mono2<-melt(mononucleotides, id.vars = c("species"), measure.vars = mono_col, variable.name = "nucleotide", value.name = "count" )
freq_mono2[, frecuencies := count/sum(count), by=.(species)]; freq_mono2

freq_di2<-melt(dinucleotides, id.vars = c("species"), measure.vars = di_col, variable.name = "dinucleotide", value.name = "count" )
freq_di2[, frecuencies := count/sum(count), by=.(species)]; freq_di2

#I am going to use freq_mono2 and freq_di2 (the ones created with melt function)
```
<span style="color:red"> 3/3 pts. Great </span>


(3 pts) - Calculate GC content in each sponge and add it as a column to the dinucleotide frequency table.
```{r}
#GC-content percentage is calculated as Count(G + C)/Count(A + T + G + C) * 100%.
temp1<-freq_mono2[, .("GC_content (%)" = (count[nucleotide=="C"] + count[nucleotide=="G"])/sum(count) * 100), by=species]
freq_di2<-merge(freq_di2, temp1, by="species"); freq_di2
```

<span style="color:green"> 3/3 Nice usage of the merge :D  </span>


(1+1+1+1 pts) - If all dinucleotides are equally probable, what is the probability for any of them = expected frequency? Which dinucleotide is underrepresented? Which are overrepresented? Are they reproducibly over/underrepresented in all sponges?
```{r}
#Probability of any of them = expected frequency
e_freq<-(1/16); e_freq

#Which dinucleotide is overrepresented --> I do not know if you wanted just the dinucleotide or dinucleotide + species, so here is both options:
var2c1<-freq_di2[frecuencies>e_freq, .(species), by=dinucleotide][order(species)]; var2c1
var2c11<-freq_di2[frecuencies>e_freq, unique(dinucleotide)];var2c11

#Which dinucleotide is underrepresented --> I do not know if you wanted just the dinucleotide or dinucleotide + species, so here is both options:
var2c2<-freq_di2[frecuencies<e_freq, .(species), by=dinucleotide]; var2c2
var2c21<-freq_di2[frecuencies<e_freq, unique(dinucleotide)]; var2c21

#In all sponges?
var2c1[order(dinucleotide), .N, by=dinucleotide][N==7] #this are the only ones that are overrepresented in all sponges
var2c2[order(dinucleotide), .N, by=dinucleotide][N==7] #this are the only ones that are underrepresented in all sponges

#Just some of the nucleotides are under/overrepresented in all sponges. The overrepresented are AA, AT, CA, TG and TT, whereas the underrepresented in all species are CC, CG, GC and GG (those related to GC content).
```
<span style="color:green"> 4/4 </span>


(5 points) - Since GC content differs between sponge species (not all mononucleotides are equally represented), the expected frequency for dinucleotides will not be the same for each sponge species. Calculate the expected frequency for each dinucleotide in each sponge species (*from mononucleotide table*) and add it to a column in dinucleotide frequency table.
```{r}
subseting<-function(x){
  dinuc<-unlist(strsplit(x, split = ""))
  tab<-freq_mono2[, .(new_freq = frecuencies[nucleotide==dinuc[1]] * frecuencies[nucleotide==dinuc[2]],
                      dinucleotide = x), by=species]
  tab
}


di<-as.character(freq_di2[, unique(dinucleotide)])
all_tables<-lapply(di, subseting)
newtable<-rbindlist(all_tables)
var2d<-merge(freq_di2, newtable, by=c("species", "dinucleotide")); var2d
```
<span style="color:green"> 5/5. </span>


(1+1+1+2 points) - Calculate (observed-expected)/expected score for each dinucleotide and each species. Which dinucleotides are overrepresented and which are underrepresented now when you have taken GC content into account? Is the over/underrepresentation reproducible for all sponge species? How would you explain it with biology in mind?
```{r}
#score for each dinucleotide and species
var2d[, score := (frecuencies - new_freq) / new_freq]; var2d

#dinucleotides overrepresented and underrepresented

var2e11<-var2d[score>0, .(species), by=dinucleotide]; var2e11 #overrepresented
#var2e12<-var2d[score>0, unique(dinucleotide)]; var2e12

var2e21<-var2d[score<0, .(species), by=dinucleotide]; var2e21
#var2e22<-var2d[score<0, unique(dinucleotide)]; var2e22

#over/underrepresentation reproducible for all species
var2e11[order(dinucleotide), .N, by=dinucleotide][N==7] #this are the only ones that are overrepresented in all sponges
var2e21[order(dinucleotide), .N, by=dinucleotide][N==7] #this are the only ones that are underrepresented in all sponges

#Now, taking everything into consideration we see that just CA and TG nucleotides are overrepresented in all sponges, whereas CC and TA are underrepresented in all sponges.

#Explanation in biology mind
#CG and TA are underrepresented in all sponges as in biology C and G do not tend to go after each other and the same for T and A. So that would explain that CA and TG dinucleotides are overrepresented. As C and G are bases that pair together (as so T and A), if they are located together in the same strand it may be possible that more mutations are produced by paring of this bases in the same string maybe forming loops or hairpins in the DNA molecule.

#GC content (info about DNA base composition) is important because is associated with genome size and holocentric chromosomal structure, also it has many importance in evolution. High GC content can be asociated to high rate of mutations and meiotic and mitotic recombination

```
<span style="color:red"> 5/5 </span>


# Task 3. Mutational signatures in human cancer (15 pts)

## Short introduction
Mutational signatures are characteristic combinations of mutation types arising from specific mutagenesis processes such as DNA replication infidelity, exogenous and endogenous genotoxins exposures, defective DNA repair pathways and DNA enzymatic editing. Type of mutations that occur are base subsitutions, indels and structural variants. 
Single base substitutions (SBS) are a replacement of a certain nucleotide base with another base. Considering the pyrimidines of the Watson-Crick base pairs, there are only six different possible substitutions: C>A, C>G, C>T, T>A, T>C, and T>G. Current SBS signatures have been identified using 96 different contexts, considering not only the mutated base, but also the bases immediately 5’ and 3’. Detailed examples of those SBS signatures can be found at https://cancer.sanger.ac.uk/cosmic/signatures/SBS/. 

## Task
Your job is to analyze mutational signatures for patients with breast adenocarcinoma (**Breast-AdenoCA**). However, your lab colleagues have not given you one tidy table for breast cancer patients for analysis but many smaller tables containing the information for all the patients which you have to combine together into one to be able to perform the analysis. 

The files at your disposal are the following:

+ MAIN TABLES containing the results from the program SigAnalyzer for mutational signatures SBS2 and SBS13 (each in a separate files, ***SBS2.RDS*** and ***SBS13.RDS***). The names of the columns and rows are missing, but you have been informed that colnames should be the name of the patients and rownames the genomic intervals. Values represent the probability of the mutational signature in specific genomic interval for each patient.
+ Table with ordered patients names. (***patients.txt***)
+ Table with ordered used genomic intervals for the SigAnalyzer. Some regions were filtered from the analysis due to low mappability. (***intervals.txt***)
+ Table with all the genomic intervals and special assigned names by your colleague. (***interval_names.txt***)

The analysis was performed on the pre-selected genomic intervals of 1 Mb (***intervals.txt***) and those intervals are the ones that should be added to the combined table of files ***SBS2.RDS*** and ***SBS13.RDS***. However, your colleague wants to have the special names of the genomic intervals in that file, the ones found under the column **interval** in the file ***interval_names.txt***. 

Your tasks are the following:

+ First, make one combined table from the tables containing results for SBS2 and SBS13 (***SBS2.RDS*** and ***SBS13.RDS***). Secondly, add a column "signature_mutations" containing the "SBS2" and "SB13" for the precise rows in the now combined (MAIN TABLE) table. (Must be done in this order for 2 pts)
```{r}
#1st: combined table (MAIN_TABLE) with results for SBS2 and SBS13
MAIN_TABLE<-rbindlist(lapply(c("../Task3/SBS2.RDS", "../Task3/SBS13.RDS"), readRDS), idcol = T); MAIN_TABLE

#2nd: "signature_mutations" column with "SBS2" and "SBS13"
MAIN_TABLE[, signature_mutations := ifelse(.id==1, "SBS2", "SBS13")]; MAIN_TABLE

```
<span style="color:green"> 2/2  </span>


+ Add patients name and special assigned genomic interval name by your colleague to the MAIN TABLE (8 pts) (Hints: Look at the both tables containing intervals. What is the difference between them? Think of your own unique way how can you fix that.). Work as tidy as you can in this task!
```{r}
#1. Addition of genomic intervals for rownames:
used_intervals<-fread("../Task3/intervals.txt")
all_intervals<-fread("../Task3/interval_names.txt")

  #Changes in tables to do merge
used_intervals[, chr := paste("chr", chr, sep="")]
names(used_intervals)[1]<-"seqnames"

  #Merging tables to obtain special assigned genomic interval names for used intervals
clean_intervals<-merge(used_intervals, all_intervals, by=c("seqnames", "start", "end"), idcol=T)

  #As r does not allow duplicated rownames (function rownames) I assign the names as identificators (.id).
MAIN_TABLE[, .id := rep(clean_intervals[, interval], times=2)]; MAIN_TABLE

#2. Patients names as colnames
patient_names<-unlist(fread("../Task3/patients.txt")); patient_names
colnames(MAIN_TABLE)[-c(1, 82)]<-patient_names

MAIN_TABLE
```
<span style="color:red"> 6/8  You have lost the proper ordering when doing merge. You had to use merge(., sort=FALSE) or some other way of correcting for the order. </span>


+ From the file=**patientsCancerType.txt** add the appropriate cancer type for each patients (3 pts)
```{r}
cancer_type<-fread("../Task3/patientsCancerType.txt")

#reshaping cancer_type table to match MAINTABLE. I will be using patient_names variable from previous task to order the cancer_type table.
cancer_type<-as.data.table(t(cancer_type))
colnames(cancer_type)<-unlist(cancer_type[1])
setcolorder(cancer_type, patient_names)
cancer_type<-cancer_type[-1, ]
cancer_type

#addtion of a new row with the cancer type for each patient.
MAIN_TABLE<-rbind(MAIN_TABLE, cancer_type, fill=TRUE)
```
<span style="color:red"> 1.75/3 You were asked to work as tidy as possible. Is adding a new row containing different type of data to the columns tidy. Your second way also produces chr type and you are missing the values. </span>

+ How many breast adenocarcinoma patients do you have and which one has the total highest probability for mutational signatures SBS2 and SBS13, respectively? (2 pts)
```{r}
#How many adenocarcinoma patients.
sum(ifelse(MAIN_TABLE[4257, .SD=="Breast-AdenoCA", .SDcols = (1:82), by=.SD], 1, 0)) # PS: The output of this command is NA because the .di and signature_mutations columns are also taken into the account. 
#result-->3

#Total highest probability for mutational signatures SBS2 and SBS13
col_numb<-which(MAIN_TABLE[4257, .SD=="Breast-AdenoCA", .SDcols = (1:82), by=.SD] == TRUE)

adeno_patients<-colnames(MAIN_TABLE[, .SD, .SDcols = c(col_numb)]); adeno_patients

last_results<-MAIN_TABLE[, .SD, .SDcols = c(col_numb), by=signature_mutations][, lapply(.SD, as.numeric), .SDcols = c(adeno_patients), by=signature_mutations][, lapply(.SD, sum), .SDcols = c(adeno_patients), by=signature_mutations]
last_results

#the one that has the total highest probability for mutational signatures SBS2 is patient YDWUUPAg and the one that has the highest probability for mutational signatures SBS13 is patient 9701KJid.
```
<span style="color:red"> 1/2  You are also getting NA row in your final table. Here you can see that the code and the overall results are heavy affected by the look of your table gained in the previous task. You lost the points there, however in this pat you have not accounted for the NAs that occur in your code. </span>
